domain/resource/state/resource.go

```diff
@@ -1356,6 +1481,542 @@ WHERE uuid = $updatePotentialResource.uuid
+FROM   charm c
+JOIN   charm_source cs ON c.source_id = cs.id
+AND    uuid = $getCharmSource.uuid
+`, source)
+	if err != nil {
+		return false, errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, getCharmSourceStmt, source).Get(&source)
+	if err != nil {
+		return false, errors.Capture(err)
+	}
+
+	return charm.Local.Matches(source.SourceName), nil
+}
+
+// getCharmResourceKind fetches the kind of charm resource and returns
+// [resourceerrors.ResourceNotFound] if it cannot be found.
+func (st *State) getCharmResourceKind(
+	ctx context.Context,
+	tx *sqlair.TX,
+	charmID charm.ID,
+	resourceName string,
+) (charmresource.Type, error) {
+	charmRes := charmResource{
+		CharmUUID:    charmID.String(),
+		ResourceName: resourceName,
+	}
+	checkCharmResourceExistsStmt, err := st.Prepare(`
+SELECT &charmResource.kind
+FROM   v_charm_resource
+WHERE  name       = $charmResource.name
+AND    charm_uuid = $charmResource.charm_uuid
+`, charmRes)
+	if err != nil {
+		return 0, errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, checkCharmResourceExistsStmt, charmRes).Get(&charmRes)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return 0, resourceerrors.ResourceNotFound
+	} else if err != nil {
+		return 0, errors.Capture(err)
+	}
+
+	kind, err := charmresource.ParseType(charmRes.Kind)
+	if err != nil {
+		return 0, errors.Errorf("parsing resource type %q: %w", charmRes.Kind, err)
+	}
+	return kind, nil
+}
+
+// getAppplicationAndCharmUUID returns gets the application ID and charm UUID
+// for the given application name, returning [resourcerrors.ApplicationNotFound]
+// if it cannot be found.
+func (st *State) getApplicationAndCharmUUID(
+	ctx context.Context,
+	tx *sqlair.TX,
+	applicationName string,
+) (application.ID, charm.ID, error) {
+	app := getApplicationAndCharmID{Name: applicationName}
+	queryApplicationStmt, err := st.Prepare(`
+SELECT (charm_uuid, uuid) AS (&getApplicationAndCharmID.*)
+FROM   application
+WHERE  name = $getApplicationAndCharmID.name
+`, app)
+	if err != nil {
+		return "", "", errors.Capture(err)
+	}
+	err = tx.Query(ctx, queryApplicationStmt, app).Get(&app)
+	if err != nil {
+		if !errors.Is(err, sqlair.ErrNoRows) {
+			return "", "", errors.Capture(err)
+		}
+		return "", "", errors.Errorf("%w: %s", resourceerrors.ApplicationNotFound, applicationName)
+	}
+	return app.ApplicationID, app.CharmID, nil
+
+}
+
+// getOriginIDs returns the database IDs for the origin types.
+func (st *State) getOriginIDs(ctx context.Context, tx *sqlair.TX) (map[string]int, error) {
+	type origin struct {
+		Name string `db:"name"`
+		ID   int    `db:"id"`
+	}
+
+	selectOriginStmt, err := st.Prepare(`
+SELECT &origin.*
+FROM   resource_origin_type
+`, origin{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var origins []origin
+	err = tx.Query(ctx, selectOriginStmt).GetAll(&origins)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	m := make(map[string]int)
+	for _, o := range origins {
+		m[o.Name] = o.ID
```

Comment:
Can they collide?

Reply:
no, the names must be unique in the table.